kyo-prelude/shared/src/main/scala/kyo/debug/Debug.scala

```diff
@@ -0,0 +1,51 @@
+package kyo
+
+import Ansi.*
+import kyo.kernel.Effect
+import kyo.kernel.Safepoint
+
+object Debug:
+
+    def apply[A, S](v: => A < S)(using frame: Frame): A < S =
+        Effect.catching {
+            v.map { value =>
+                println(frame.show)
+                printValue(value)
+                value
+            }
+        } { ex =>
+            println(frame.show)
+            printValue(ex)
+            throw ex
+        }
+    end apply
+
+    def trace[A, S](v: => A < S): A < S =
+        val interceptor = new Safepoint.Interceptor:
+            def enter(frame: Frame, value: Any): Boolean =
+                printValue(value)
+                println(frame.parse.show)
+                true
+            end enter
+            def addEnsure(f: () => Unit): Unit    = ()
+            def removeEnsure(f: () => Unit): Unit = ()
+
+        Safepoint.propagating(interceptor) {
+            Effect.catching {
+                v.map { value =>
+                    printValue(value)
+                    value
+                }
+            } { ex =>
+                printValue(ex)
+                throw ex
+            }
+        }
+    end trace
+
+    private def printValue(value: Any) =
```

Comment:
there seems like an opportunity to print the frame as well, but perhaps not all the usages need that.

Reply:
Both callers of this method print the frame but in different order. `Debug.apply` logs the frame and then the value since it's the result of the transformation but `Debug.trace` prints the value before the frame because it's the input.